spring-integration-mail/src/test/java/org/springframework/integration/mail/ImapMailReceiverTests.java

```diff
@@ -1063,6 +1067,21 @@ public void writeTo(OutputStream os) throws IOException, MessagingException {
 			}
 			super.writeTo(os);
 		}
+
+		@Override
+		public synchronized void setFlags(Flags flag, boolean set) throws MessagingException {
```

Comment:
Why do we need to override this if there is that `Message.getFlags()`?

Reply:
This is to show that we are calling the `setFlags` on the original message and not the copied one. The `Flags` can be shared. The `MimeMessage` has

```java
flags = source.getFlags();
if (flags == null)    // make sure flags is always set
    flags = new Flags();
```

So if the source has `Flags` and we call the `setFlags` on the `filteredMessages` then the test will still pass. The goal of the change is to show that the `setFlags` is called on the original message.

Reply:
Doesn't look like you validate somehow in the test if message was original or `IntegrationMimeMessage`.
Therefore this extra precaution is redundant.

Reply:
There is

```java
assertThat(msg2.getMyTestFlags().contains(Flag.SEEN)).isTrue();
```

In the test.

Reply:
But looks like it is exactly the same as previous one:
```
assertThat(msg2.getFlags().contains(Flag.SEEN)).isTrue();
```

That's why I believe that is really enough.
Because you are not going to deal with mock messages in the real application.

Reply:
I added this to validate that the `setFlags` is called on the original message and not on the `IntegrationMimeMessage`. Currently, `GreenMailUtil#newMimeMessage` does not set the flags. However, if you want I can set it up in such case that if the `setMessageFlagsAndMaybeDeleteMessages` is called on the `IntegrationMimeMessage` the test would fail.

```java
assertThat(msg2.getFlags().contains(Flag.SEEN)).isTrue();
```

The test above passes as the `Flags` are shared. But

```java
assertThat(msg2.getMyTestFlags().contains(Flag.SEEN)).isTrue();
```

would fail as it has it's own flags. If you think it is too much, I can remove it

Reply:
OK. You probably can `spy()` on the `GreenMailUtil#newMimeMessage` result to verfiy that exactly its `setFlags()` is called.
It is still look redundant to have our own property when it is there in the super class.

Reply:
good point with the `spy()`. I'll add that

Reply:
I've replaced that with a spy. Hope it is good now.